# TEdetectionEvaluation

Tool comparison for detecting differentially expressed individual transposable
elements.

This evaluation is based on raw count tables. It supports raw count tables
generated by salmonTE, SQuIRE, TEtools, Telescope and TEtranscripts.


## Run evaluation

The path of the directories where the results are located, the count tables (in
case of SQuIRE the common prefix), and addition files files have to sign in into
the `dataInfo.csv`. The evaluation process needs a reference to compare the
results of the tools. These reference has to be stored under Simulation.

Additional Files:

-   SQuIRE needs a 'dictionary' to translate TE ids. This file can be generated
    with `generateDict.py` and will be explained further down.

-   TEtools needs a file where the order of the fastqs are listed (order of the
    original TEtools call) without the extension `.fastq`, e.g.:

        sample_1
        sample_2
        sample_3
          .
          .
          .

When the data is filled in run `Rscript TEdetectEval.R` to run the evaluation
procedure. Subsequently, by running `Rscript figures.R` and `Rscript tables.R`
the figures and tables were generated.

# Helper Scripts

## Generation of SQuIRE dictionary

SQuIRE hast a method to generate a .bed-file where a TE identifier is used which
identifies each instance also in the resulting count table. The identifier for
each TE is in following format:

`chr|start|end|TE-subfamily:TE-family:TE-repclass|score|strand`

However, the TE identifier that that is used in the simulated data set is
assembled as following:

`chr|start|end|TE-repclass|TE-family|TE-subfamily|score|Kimura distance`

For the evaluation it is necessary which simulated TE is detected by SQuIRE so
that a dictionary is generated to translate the TE ids. Since chr, start and end
are unique for each TE these three values are used to translate the ids and to
get a table for the TE ids that belonging together.

This can be done with the helper script `generateDict.py`, which needs as input
the bed-file of SQuIRE and your own.

## Generate TE reference library

An .align-file generated by RepeatMasker is needed to generate such library. The
helper script can be used to generate the reference library. Besides of the
align file the reference genome in fasta format is also needed.

`bash alignToFasta.sh <.align-file> <referenceGenome.fa>`
